packages/services/src/kernel/default.ts

```diff
@@ -1527,11 +1532,13 @@ namespace Private {
     // We might want to move the handleRestart to after we get the response back
 
     // Handle the restart on all of the kernels with the same id.
-    each(runningKernels, k => {
-      if (k.id === kernel.id) {
-        k.handleRestart();
-      }
-    });
+    await Promise.all(
+      runningKernels.map(async k => {
+        if (k.id === kernel.id) {
+          await k.handleRestart();
```

Comment:
I think what you want here is `return k.handleRestart()` and not `await`. And I think the iterator function should not be `async`.

Reply:
Aren't they equivalent?

Reply:
That's a good question. Let's compare these two snippets and their generated JS:

https://www.typescriptlang.org/play/#src=(async%20()%20%3D%3E%20%7B%0D%0A%20%20const%20generator%20%3D%20n%20%3D%3E%20%7B%0D%0A%20%20%20%20console.log('generating%20promise%20for'%2C%20n)%3B%0D%0A%20%20%20%20return%20Promise.resolve(n)%3B%0D%0A%20%20%7D%3B%0D%0A%20%20return%20Promise.all(%5B1%2C%202%2C%203%5D.map(async%20n%20%3D%3E%20await%20generator(n)))%3B%0D%0A%7D)().then(result%20%3D%3E%20%7B%0D%0A%20%20console.log('result%20is'%2C%20result)%3B%0D%0A%7D)%3B

vs.

https://www.typescriptlang.org/play/#src=(async%20()%20%3D%3E%20%7B%0D%0A%20%20const%20generator%20%3D%20n%20%3D%3E%20%7B%0D%0A%20%20%20%20console.log('generating%20promise%20for'%2C%20n)%3B%0D%0A%20%20%20%20return%20Promise.resolve(n)%3B%0D%0A%20%20%7D%3B%0D%0A%20%20return%20Promise.all(%5B1%2C%202%2C%203%5D.map(n%20%3D%3E%20generator(n)))%3B%0D%0A%7D)().then(result%20%3D%3E%20%7B%0D%0A%20%20console.log('result%20is'%2C%20result)%3B%0D%0A%7D)%3B

They are definitely *different*, but I'm not sure whether they are logically different and doing more work. I *think* it's the case that my suggestion above has one fewer indirection. What do you think?

Reply:
To me they both look like equivalent asynchronous functions that return `Promise<void>` (at least when using ES6). I have no objection to your version, just trying to understand if there is a logical difference.

Reply:
How about now?

Reply:
Does returning a promise from an async function resolve that or return it nested? That seems to be the underlying question... For consistency I would think it would be nested, but Typescript seems to think it awaits it...

Reply:
I believe it returns the promise. cf. this discussion: https://eslint.org/docs/rules/no-return-await

Reply:
Right, I think that the compiled JS output is always wrapping the return, even when the return is a promise.

Reply:
Ah I see. I misunderstood how [`Promise.resolve`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve) worked. I thought it unconditionally wrapped it's argument in a promise. But actually:

> If the value is a promise, that promise is returned;

This doesn't seem very composable! What if I wanna have a promise of a promise??